### DISCUSSION TITLE
How will new validations be decided?

### DISCUSSION THREAD

**Discussion by louiss0** [open-circle/valibot#53] [Q&A] [answered]
I'm the kind of person that likes to ask for new features that I believe are beneficial to developers. 
But I have noticed that it's difficult to get maintainers to add features to libraries even though it's beneficial and would be 
appreciated. 

If I wanted to ask for a new transformation or validation how would you decide whether it's worth adding?

**Comment by fabian-hiller** [comment, marked as answer]
If it is generally useful, needed by many or corresponds to a standard, for example like ISO timestamp or UUID, I am open for any ideas. Due to the modular design it has no impact on existing users if we extend the functionality here. 

A validation of credit cards would be cool for example. I just added a function today with [`isLuhnAlgo`](https://github.com/fabian-hiller/valibot/blob/main/library/src/utils/isLuhnAlgo/isLuhnAlgo.ts) that could be useful for this. On the other hand, I'm cautious about things like phone number, as it will be impossible to cover every country and variant.

**Reply by louiss0** [reply]
Do you happen to have a validator that validates characters in a string but ignores spaces? I think that's useful and be the default. What if there was a validator that validated characters based on punctuation?

**Reply by fabian-hiller** [reply]
Do you have a concrete example how code with such a validation looks like? Basically anything can be validated using `regex`, `custom` or a custom function. Valibot does not limit this.

**Comment by louiss0** [comment]
> Do you have a concrete example how code with such a validation looks like? Basically anything can be validated using `regex`, `custom` or a custom function. Valibot does not limit this.


I'm sorry I don't know the library so I'll have to use a regular function. 

Characters based on only letters not spaces 

```ts
 
function numOfLettersInThisStringIsEqualToTheNumber(string:string, number:number) {

const  trimmedStringLength = string.trim().length 

if(trimmedStringLength  !== number)

throw new Error("The characters in this string are invalid")
}

```

**Reply by fabian-hiller** [reply]
You can already do this easily with Valibot:

```ts
import { email, endsWith, string, toTrimmed } from 'valibot'; // 0.57 kB

const EmailSchema = string([toTrimmed(), email(), endsWith('@example.com')]);
```

More info here: https://valibot.dev/guides/pipelines/

**Reply by louiss0** [reply]
I was looking for the number of letters in the string in that example. 

I think you meant

```js
import { size string, toTrimmed } from 'valibot'; // 0.57 kB

const numOfCharsSchema = string([toTrimmed(), size(60) ]);

```

Either way letters is not enough punctuation would also have to be taken to account.  

The user would have to I think write this if I'm correct.

```js
import { size() string, toTrimmed } from 'valibot'; // 0.57 kB

const numOfCharsSchema = string([toTrimmed(),   size()]);
```

**Reply by fabian-hiller** [reply]
For the validation of a length Valibot provides `length`, `minLength` and `maxLength`. Note that `toTrimmed` will change the string. If you don't want this, you can use `custom` to write your own validation logic.

```ts
import { custom, length, string, toTrimmed } from 'valibot';

const numOfCharsSchema1 = string([toTrimmed(), length(60)]);
const numOfCharsSchema2 = string([custom((input) => input.trim().length === 60)]);
```

More info here: https://valibot.dev/guides/pipelines/

**Reply by louiss0** [reply]
Got it

**Comment by louiss0** [comment]
> For the validation of a length Valibot provides `length`, `minLength` and `maxLength`. Note that `toTrimmed` will change the string. If you don't want this, you can use `custom` to write your own validation logic.
> 
> ```ts
> import { custom, length, string, toTrimmed } from 'valibot';
> 
> const numOfCharsSchema1 = string([toTrimmed(), length(60)]);
> const numOfCharsSchema2 = string([custom((input) => input.trim().length === 60)]);
> ```
> 
> More info here: https://valibot.dev/guides/pipelines/
